### PR TITLE
⏱️ Uptime rewards and phase events

### PR DESCRIPTION
**Pull request by bdeneux**
This PR add detection of phases event like when a new phase start or end. Also by detecting if a phase is finish, a new mongo query has been added to calculate the uptime point based on the task description.

#### 📥 Events 

I call this event but not event has been created, it's more like a detection of event on each new block. If a block is the first block of a phase, this trigger a function to do some action and the same for the last block of a phase. 

No action has been configured on first phase block but a new action has been set on the end of a phase : calculate uptime. 

#### ⏱️ Uptime
The calcul used for uptime calculation is here : 
`2501^0.01x - 1` with x = uptime percentage. That's give in mongo : 

```json
{
   "uptime":{
      "$subtract":[
         {
            "$pow":[
               2501,
               {
                  "$multiply":[
                     0.01,
                     {
                        "$subtract":[
                           100,
                           {
                              "$divide":[
                                 {
                                    "$multiply":[
                                       100,
                                       "$totalMissedBlock"
                                    ]
                                 },
                                 {
                                    "$subtract":[
                                       "$phase.blocks.to",
                                       "$phase.blocks.from"
                                    ]
                                 }
                              ]
                           }
                        ]
                     }
                  ]
               }
            ]
         },
         1
      ]
   }
}
```

The mongo aggregation will before calulate uptime, detecting all missedBlocks in the given phase and cut missedBlocks range if a missed block range is between two phases. 
https://github.com/okp4/nemeton-leaderboard/blob/d87dac24d2c763ec630b289fd08a7aa610881707/app/nemeton/store.go#L479-L508

### CODE REVIEW

**Comment by bdeneux** [comment]
@amimart Since reward point is in Long format, I've fixed mongo aggregation to make an absolute round on points calculation when setting points on validator collection. I've decide to cut the decimal number since a validator with 2499,6 points cannot be rounded to 2500 points because he missed some block. 
So I've used `{ "$toLong": "$$newPoints" }` to convert 2499,6 to 2499 points. What do you think about that ? Or maybe do we need to made a correct rounding?

File: app/nemeton/store.go [hunk 1]
```diff
@@ -433,3 +433,127 @@ func (s *Store) GetPhaseBlocks(ctx context.Context, number int) (*BlockRange, er
...
+		bson.M{"$group": bson.M{
+			"_id":              "$validator",
+			"totalMissedBlock": bson.M{"$sum": "$count"},
+			"phase":            bson.M{"$mergeObjects": "$currentPhase"},
+			"points":           bson.M{"$first": "$points"},
+		}},
+		bson.M{"$addFields": bson.M{
+			"uptime": bson.M{"$subtract": bson.A{
+				bson.M{
+					"$pow": bson.A{2501, bson.M{
```
Comment (amimart) [inline app/nemeton/store.go:523]: The uptime won't attribute the same number of points every phase, parameters have been introduced in the task structure so you can retrieve this value with: `task.GetUptimeMaxPoints() + 1`

**Review by amimart — CHANGES_REQUESTED** [review]
Awesome 😍 

Just a remarks regarding the maximum number of points the uptime task can attribute, which can vary.

File: app/nemeton/store.go (hunk 1 above)
Comment (bdeneux) [reply app/nemeton/store.go:523]: Ah yes ! I've seen that has been introduce in the last PR #43, I will update this.

**Review by amimart — APPROVED** [review]
Let's go!

**Pull request merged**